src/nupic/algorithms/Anomaly.cpp

```diff
@@ -29,33 +29,49 @@
 #include "nupic/algorithms/Anomaly.hpp"
 #include "nupic/utils/Log.hpp"
 #include "nupic/utils/MovingAverage.hpp"
+#include "nupic/types/SdrTools.hpp" // sdr::Intersection
 
 using namespace std;
+using namespace nupic;
+using namespace nupic::algorithms::anomaly;
+using namespace nupic::sdr;
 
 namespace nupic {
 namespace algorithms {
 namespace anomaly {
+Real32 computeRawAnomalyScore(SDR& active,
+                              SDR& predicted) {
 
-
-Real computeRawAnomalyScore(const vector<UInt>& active,
-                              const vector<UInt>& predicted)
-{
   // Return 0 if no active columns are present
-  if (active.size() == 0) {
+  if (active.getSum() == 0) {
     return 0.0f;
   }
 
-  set<UInt> active_{active.begin(), active.end()};
-  set<UInt> predicted_{predicted.begin(), predicted.end()};
-  vector<UInt> predictedActiveCols;
+  NTA_CHECK(active.dimensions == predicted.dimensions);
 
   // Calculate and return percent of active columns that were not predicted.
-  set_intersection(active_.begin(), active_.end(), predicted_.begin(),
-                   predicted_.end(), back_inserter(predictedActiveCols));
+  Intersection both(active, predicted); //TODO allow Intersection to act on const SDR, so active,pred can be const
+
+  cout << "act " << active << endl;
+  cout << "pred " << predicted << endl;
+  cout << "x " << both;
+
+  return (active.getSum() - both.getSum()) / Real(active.getSum());
+}
+
+Real32 computeRawAnomalyScore(const vector<UInt>& active,
+                              const vector<UInt>& pred) {
+  SDR a({static_cast<UInt>(active.size())});
+  a.setSparse(active);
```

Comment:
The problem is here.  I don't know what data format active is, but either way it's incorrect.  The also applies to the predicted SDR.
* If active is a list of sparse indexes then active.size() is not the total size of the SDR.
* If active is a dense array of boolean values, then this needs to call setDense, not setSparse. 

Compiling with `cmake ../.. -DCMAKE_BUILD_TYPE=Debug` will catch this kind of error.

Reply:
It might be easier to write this the other way around, to have the SDR API of this code call the old vectorAPI.  Something to consider?

Reply:
> If active is a list of sparse indexes then active.size() is not the total size of the SDR. 

thanks! Yes, the vectors should be sparse. 

Thanks for the fixes, I'll just revert the implementation for Anomaly::compute(vector, vector), as the aim is to call just one implementation (SDR based)